feat: coalesced event

Gregoirevda opened the pull request:
## Summary
`onInsetsChange` can sometimes be sent very quickly after another. Using the same technique as the ScrollView, we keep the view property `onInsetsChange`, but it's triggered from a custom `RCTEvent` with a coalescingKey.

If events are sent 5 times very quickly, only 1 event will be sent with the latest insets and frame.

Fixes #579 #485 #172 

Note:
I've tested this in a real app and `onInsetsChange` calls goes from 3 to 1, but in the example app it doesn't change (still 3 calls when rotating) and that's because it doesn't change anything in batching which is good. 
So, if during a flush to JS 5 `onInsetsChange` are sent only the last one will survive, but if the app isn't busy and it can send the 5 events independently it will.

`EventEmitter` will collect the event in a queue and then flush them. Coalesce changes a flush with multiple events from
`[onInsetsChange x1 y1, onInsetsChange x2 y2]` to
`[onInsetsChange x2 y2]`


PS:
I've only implemented the old architecte, if this can be reviewed I can then only apply changes to this part and finish the fabric arch.
Requires `RCT_NEW_ARCH_ENABLED=0 pod install` to test.

jacobp100 commented on ios/RNCSafeAreaProvider.m:
```diff
@@ -69,27 +84,23 @@ - (void)invalidateSafeAreaInsets
     return;
   }
-
+  
```
```suggestion

```

jacobp100 commented on ios/RNCSafeAreaProvider.m:
```diff
@@ -99,4 +110,13 @@ - (void)layoutSubviews
   [self invalidateSafeAreaInsets];
 }
 
+- (void)dealloc
```
Minor - could you put this dealloc function after the `init` functions

jacobp100 commented on src/SafeAreaContext.tsx:
```diff
@@ -58,6 +58,8 @@ export function SafeAreaProvider({
       nativeEvent: { frame: nextFrame, insets: nextInsets },
     } = event;
 
+    console.log('onInsetsChange', nextInsets);
+
```
```suggestion
```

jacobp100 commented on ios/RNCSafeAreaProvider.m:
```diff
@@ -1,19 +1,33 @@
   UIEdgeInsets _currentSafeAreaInsets;
   CGRect _currentFrame;
   BOOL _initialInsetsSent;
+  uint16_t _coalescingKey;
```
I can't see how this would ever be changed from the default of `0` - is it actually needed?

jacobp100 commented on ios/RNCSafeAreaProvider.m:
```diff
@@ -99,4 +106,12 @@ - (void)layoutSubviews
   [self invalidateSafeAreaInsets];
 }
 
+RCT_NOT_IMPLEMENTED(-(instancetype)initWithFrame : (CGRect)frame)
+RCT_NOT_IMPLEMENTED(-(instancetype)initWithCoder : (NSCoder *)aDecoder)
```
Minor

```suggestion
RCT_NOT_IMPLEMENTED(-(instancetype)initWithFrame:(CGRect)frame)
RCT_NOT_IMPLEMENTED(-(instancetype)initWithCoder:(NSCoder *)aDecoder)
```

jacobp100 reviewed (approved):
Thank you for this!

Interesting it didn't fully resolve the issue. But at least it's not worse

jacobp100 commented:
No - if you add `[_eventDispatcher.bridge.uiManager.observerCoordinator addObserver:self];` and `- (void) uiManagerDidPerformLayout`, the latter function will get called every time RN updates the layout

Gregoirevda commented:
@jacobp100 I've set a flag in `invalidateSafeAreaInsets` and added the observer in `init` and only emit the event in `uiManagerDidPerformLayout` and this won't work:

- **Initial layout:**
`uiManagerDidPerformLayout` isn't called after the component mounted. I've fixed this by emitting in the `invalidateSafeAreaInsets` when `_initialInsetsSent` hasn't been set.

- **portrait to landscape** event order:
```
uiManagerWillPerform
uiManagerDidPerform
invalidateSafeAreaInsets
```
=> no events sent

- **landscape to portait** event order
```

invalidateSafeAreaInsets
uiManagerWillPerform
uiManagerDidPerform
uiManagerWillPerform
uiManagerDidPerform
..... many times
invalidateSafeAreaInsets (stops here)
```
=> I can toggle the flag once the event is emitted in `uiManagerDidPerform`, but the last event wont be sent

Gregoirevda commented:
@janicduplessis updated with clang format
